services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java

```diff
@@ -0,0 +1,105 @@
+
+import org.jboss.logging.Logger;
+import org.keycloak.Token;
+import org.keycloak.TokenCategory;
+import org.keycloak.crypto.Algorithm;
+import org.keycloak.crypto.KeyUse;
+import org.keycloak.crypto.SignatureProvider;
+import org.keycloak.crypto.SignatureSignerContext;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.TokenManager;
+import org.keycloak.protocol.oidc.OIDCConfigAttributes;
+
+public class DefaultTokenManager implements TokenManager {
+
+    private static final Logger logger = Logger.getLogger(DefaultTokenManager.class);
+
+    private static String DEFAULT_ALGORITHM_NAME = Algorithm.RS256;
+
+    private final KeycloakSession session;
+
+    public DefaultTokenManager(KeycloakSession session) {
+        this.session = session;
+    }
+
+    @Override
+    public String encode(Token token) {
+        String signatureAlgorithm = signatureAlgorithm(token.getCategory());
+
+        SignatureProvider signatureProvider = session.getProvider(SignatureProvider.class, signatureAlgorithm);
+        SignatureSignerContext signer = signatureProvider.signer();
+
+        String encodedToken = new JWSBuilder().type("JWT").jsonContent(token).sign(signer);
+        return encodedToken;
+    }
+
+    @Override
+    public <T extends Token> T decode(String token, Class<T> clazz) {
+        if (token == null) {
+            return null;
+        }
+
+        try {
+            JWSInput jws = new JWSInput(token);
+
+            String signatureAlgorithm = jws.getHeader().getAlgorithm().name();
+
+            SignatureProvider signatureProvider = session.getProvider(SignatureProvider.class, signatureAlgorithm);
+            if (signatureProvider == null) {
+                return null;
+            }
+
+            String kid = jws.getHeader().getKeyId();
+            // Backwards compatibility. Old offline tokens and cookies didn't have KID in the header
+            if (kid == null) {
+                logger.debugf("KID is null in token. Using the realm active key to verify token signature.");
+                kid = session.keys().getActiveKey(session.getContext().getRealm(), KeyUse.SIG, signatureAlgorithm).getKid();
+            }
+
+            boolean valid = signatureProvider.verifier(kid).verify(jws.getEncodedSignatureInput().getBytes("UTF-8"), jws.getSignature());
+            return valid ? jws.readJsonContent(clazz) : null;
+        } catch (Exception e) {
+            logger.debug("Failed to decode token", e);
+            return null;
+        }
+    }
+
+    @Override
+    public String signatureAlgorithm(TokenCategory category) {
+        switch (category) {
+            case INTERNAL:
+                return Algorithm.HS256;
+            case ADMIN:
+                return getSignatureAlgorithm(null);
+            case ACCESS:
+                return getSignatureAlgorithm(OIDCConfigAttributes.ACCESS_TOKEN_SIGNED_RESPONSE_ALG);
```

Comment:
Regarding access tokens, we may possibly need different signature/encryption algorithms for access tokens not just according to "frontend-client", but also based on service, which is invoked. For example "service-a" wants RS256 when "service-b" wants RS384. So frontend-client may want different access tokens algorithms based on the target service. I guess we already discussed this, but not sure what we decided on :)

This is related to audience support and the "resource" specification (I sent mail yesterday evening). Probably it's fine to have accessTokenSignedResponseAlg on cliet for now and target this somewhen in the future when needed...

Reply:
This is related to the "resource" specs I mentioned with the relation to the audience support. One cool thing about this "resource" parameter is that it is added in the code-to-token or refresh token requests. No need to re-trigger whole login flow. And it specifically mentions that accessToken may be signed/encrypted based on the "resource" (target service, which accessToken will be used against). Some cite from that specs:

```
The "resource" parameter indicates the physical location of resource
   server, typically as an https URL, where the client intends to use
   the requested access token.  This enables the authorization server to
   apply policy as appropriate for the resource, such as determining the
   type and content of the token to be issued, if and how the token is
   to be encrypted, and applying appropriate audience restrictions to
   the token.
```

+1 to address this as a future requirement. Just wanted to point and discuss a bit :)